homework/philosopher.cpp

```diff
@@ -0,0 +1,144 @@
+#include <algorithm>    // std::for_each()
+#include <atomic>
+#include <chrono>
+#include <cstdlib>
+#include <ctime>
+#include <functional>   // std::mem_fn() 
+#include <iostream>
+#include <mutex>
+#include <string>
+#include <sstream>
+#include <thread>
+
+constexpr size_t itemsCnt = 5;       // number of items (philosophers, forks)
+constexpr size_t actionCnt = 10;     // number of action to take by philosopher
+constexpr bool action_debug = false; // additional debugs
+
+std::mutex forks[ itemsCnt ];
+std::thread philosopher[itemsCnt];
+
+std::mutex displayMutex;
+std::mutex updateMutex;
+std::atomic<bool> canEat[itemsCnt];
+std::atomic<bool> eat[itemsCnt];
+
+using namespace std::chrono_literals;
+
+void displayState()
+{
+    std::stringstream ss;
+
+    std::lock_guard<std::mutex> lock( displayMutex );
+
+    for ( int i = 0; i < itemsCnt; ++i )
+    {
+        if ( eat[i] )
+        {
+            ss << "-F" << i << "- ";
+        }
+        else
+        {
+            ss << " F" << i << "  ";
+        }
+    }
+    ss << std::endl;
+    std::cout << ss.str();
+}
+
+void action( int idx )
+{
+    int left_fork = idx;
+    int right_fork = ( left_fork + 1 ) % 5;
+    int iActionStep = 0;
+
+    // std::cout << "action idx " << idx << " l " << left_fork << " r " << right_fork << std::endl; 
+
+    while ( iActionStep < actionCnt )
+    {
+        if (canEat[idx] == true)
+        {
+            std::scoped_lock lock( forks[left_fork], forks[right_fork] );
+            
+            // NOTE: Here is a race condition on eat[idx] variable and display_state.
+            // Race can be solved by using i.e. queue.
+
+            displayMutex.lock();
+            eat[idx] = true;
+            displayMutex.unlock();
```

Comment:
Na pewno displayMutex tutaj chcesz blokować? To raczej nic nie zmieni. Nie pozwalasz na wypisywanie niczego na czas modyfikacji tej zmiennej, ale wyścigi na wyświetlaniu dalej są, np. przed samą blokadą. 
Możesz spróbować do scoped_lock przekazać jeszcze mutex do wyświetlania. Wtedy tylko jeden wątek może na raz aktualizować swój stan i od razu go wyświetli. To pewnie będzie najprostsze rozwiązanie, ale przez to nie pozwalamy innym na działanie w tym samym czasie, więc nie rozwiązuje to do końca samego problemu filozofów. Ale jeśli chcesz podebugować i sprawdzić czy wyświetlanie i działanie programu jest prawidłowe to będzie ok.

Reply:
Zobacz proszę moją poprzednią uwagę.
Myślałem o tym aby do scoped_lock przekazaś jeszcze updateMutex ale wtedy stawiasz wiśksze wymagania dla blokady oraz oczywiście tylko jeden mutex w czasie może efektywnie pracować